cmd/skunk/adapter/in/networkAdapter/networkAdapter.go

```diff
@@ -0,0 +1,250 @@
+				SenderAddress:   n.peer.Address,
+				ReceiverAddress: address,
+				ChatID:          "",
+				Operation:       network.USER_OFFLINE,
+			}
+			n.SendNetworkMessageToSubscriber(message)
+			return nil
+		}
+	}
+
+	err := n.peer.SetWriteConn(address)
+	if err != nil {
+		return err
+	}
+
+	jsonMessage, err := json.Marshal(message)
+	if err != nil {
+		return err
+	}
+
+	err = n.peer.WriteMessage(string(jsonMessage))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SendNetworkMessageToSubscriber forwards a network message to the subscribed observer
+func (n *NetworkAdapter) SendNetworkMessageToSubscriber(message network.Message) {
+	n.subscriber.Notify(message)
+}
+
+// readNetworkMessages reads messages from the network and forwards them to the subscriber
+func (n *NetworkAdapter) readNetworkMessages() {
+	messageCh := make(chan string)
+	errorCh := make(chan error)
+
+	// read messages asynchronously; terminate on unsubscribe
+	// also closes messageCh and errorCh
+	go n.peer.ReadMessages(messageCh, errorCh)
+
+	// handle incoming messages and errors
+	for {
+		select {
+		case msg, ok := <-messageCh:
+			if !ok {
+				return
+			}
+			message := network.Message{}
+			err := json.Unmarshal([]byte(msg), &message)
+			if err != nil {
+				continue // optionally we could handle incorrect formatted messages
+			}
+			n.SendNetworkMessageToSubscriber(message)
+		case err, ok := <-errorCh:
+			if !ok {
+				return
+			}
+			// message subscriber that a peer is offline
+			message := network.Message{
+				Id:              util.UUID(),
+				Timestamp:       util.CurrentTimeMillis(),
+				Content:         "",
+				FromUser:        "",
+				SenderAddress:   n.peer.Address,
+				ReceiverAddress: err.Error(), // error contains the address
+				ChatID:          "",
+				Operation:       network.USER_OFFLINE,
+			}
+			n.SendNetworkMessageToSubscriber(message)
+		}
+	}
+}
+
+// startTor initializes and starts a Tor service.
+func startTor() (*tor.Tor, *cretztor.OnionService, error) {
+	conf := tor.TorConfig{
+		SocksPort:            SocksPort,
+		LocalPort:            LocalPort,
+		RemotePort:           RemotePort,
+		ReusePrivateKey:      ReusePrivateKey,
+		DeleteDataDirOnClose: DeleteDataDirOnClose,
+		UseEmbedded:          UseEmbedded,
+	}
+
+	torInstance, err := tor.NewTor(&conf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = torInstance.StartTor()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	onionService, err := torInstance.StartHiddenService()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return torInstance, onionService, nil
+}
+
+// startPeer initializes a peer network connection using the provided OnionService
+func startPeer(onionService *cretztor.OnionService) (*peer.Peer, error) {
+	peerInstance, err := peer.NewPeer(onionService.ID+".onion", LocalPort, RemotePort, "127.0.0.1:"+SocksPort)
+	if err != nil {
+		return nil, err
+	}
+	peerInstance.Listen(onionService)
+
+	time.Sleep(1 * time.Second)
```

Comment:
Is time.Sleep the best solution?

Reply:
no, i just didn't have a better solution. but now we have.